src/action/finder/request.rs

```diff
@@ -0,0 +1,1405 @@
+//
+// Use of this source code is governed by an MIT-style license that can be found
+// in the LICENSE file or at https://opensource.org/licenses/MIT.
+
+//! Defines an internal type for client side file finder action and provides
+//! a function converting proto format of the request
+//! `rrg_proto::FileFinderArgs` to the internal format.
+
+use crate::session::{
+    parse_enum, time_from_micros, ParseError, ProtoEnum, RegexParseError,
+};
+use regex::Regex;
+use rrg_proto::{
+    FileFinderAccessTimeCondition, FileFinderAction, FileFinderArgs,
+    FileFinderCondition, FileFinderContentsLiteralMatchCondition,
+    FileFinderContentsRegexMatchCondition, FileFinderDownloadActionOptions,
+    FileFinderExtFlagsCondition, FileFinderHashActionOptions,
+    FileFinderInodeChangeTimeCondition, FileFinderModificationTimeCondition,
+    FileFinderSizeCondition, FileFinderStatActionOptions,
+};
+use std::convert::TryFrom;
+
+type HashActionOversizedFilePolicy =
+    rrg_proto::file_finder_hash_action_options::OversizedFilePolicy;
+type DownloadActionOversizedFilePolicy =
+    rrg_proto::file_finder_download_action_options::OversizedFilePolicy;
+type RegexMatchMode =
+    rrg_proto::file_finder_contents_regex_match_condition::Mode;
+type LiteralMatchMode =
+    rrg_proto::file_finder_contents_literal_match_condition::Mode;
+type ActionType = rrg_proto::file_finder_action::Action;
+type ConditionType = rrg_proto::file_finder_condition::Type;
+type XDevMode = rrg_proto::file_finder_args::XDev;
+type PathType = rrg_proto::path_spec::PathType;
+
+#[derive(Debug)]
+pub struct Request {
+    pub paths: Vec<String>,
+    pub path_type: PathType,
+    pub action: Action,
+    pub conditions: Vec<Condition>,
+    pub process_non_regular_files: bool,
+    pub follow_links: bool,
+    pub xdev_mode: XDevMode,
+}
+
+#[derive(Debug)]
+pub enum Action {
+    Stat(StatActionOptions),
+    Hash(HashActionOptions),
+    Download(DownloadActionOptions),
+}
+
+#[derive(Debug)]
+pub struct StatActionOptions {
+    pub resolve_links: bool,
+    pub collect_ext_attrs: bool,
+}
+
+#[derive(Debug)]
+pub struct HashActionOptions {
+    pub max_size: u64,
+    pub oversized_file_policy: HashActionOversizedFilePolicy,
+    pub collect_ext_attrs: bool,
+}
+
+#[derive(Debug)]
+pub struct DownloadActionOptions {
+    pub max_size: u64,
+    pub oversized_file_policy: DownloadActionOversizedFilePolicy,
+    pub use_external_stores: bool,
+    pub collect_ext_attrs: bool,
+    pub chunk_size: u64,
+}
+
+#[derive(Debug)]
+pub enum Condition {
+    MinModificationTime(std::time::SystemTime),
+    MaxModificationTime(std::time::SystemTime),
+    MinAccessTime(std::time::SystemTime),
+    MaxAccessTime(std::time::SystemTime),
+    MinInodeChangeTime(std::time::SystemTime),
+    MaxInodeChangeTime(std::time::SystemTime),
+    MinSize(u64),
+    MaxSize(u64),
+    ExtFlagsLinuxBitsSet(u32),
+    ExtFlagsLinuxBitsUnset(u32),
+    ExtFlagsOsxBitsSet(u32),
+    ExtFlagsOsxBitsUnset(u32),
```

Comment:
Original file-finder groups these into one condition. Why did you decide to split these? Usually enums should be fairly small, otherwise they encourage doing wildcard pattern matching.

Reply:
I was considering it, but I believe that storing 4 Optional<Condition> is worse than adding 4 values to this enum. I want to keep those conditions as simple as possible.